File: data/derived_cot/rq3_prompting/results/translation/r1/java_to_py/code_output/DecryptionUtils.py
import math

class DecryptionUtils:
    def __init__(self, key: str):
        self.key = key

    def caesarDecipher(self, ciphertext: str, shift: int) -> str:
        plaintext = []
        shift = int(math.fmod(shift, 26))  # match Java's sign-sensitive modulo
        base_a = ord('A')
        base_a_lower = ord('a')
        for ch in ciphertext:
            if ch.isalpha():
                if ch.isupper():
                    base = base_a
                else:
                    base = base_a_lower
                shifted_char = chr((ord(ch) - base - shift + 26) % 26 + base)
                plaintext.append(shifted_char)
            else:
                plaintext.append(ch)
        return ''.join(plaintext)

    def vigenereDecipher(self, ciphertext: str) -> str:
        decrypted_text = []
        key_index = 0
        key_len = len(self.key)
        base_a = ord('A')
        base_a_lower = ord('a')
        for ch in ciphertext:
            if ch.isalpha():
                shift = ord(self.key[key_index % key_len].lower()) - base_a_lower
                if ch.isupper():
                    base = base_a
                else:
                    base = base_a_lower
                decrypted_char = chr((ord(ch) - base - shift + 26) % 26 + base)
                decrypted_text.append(decrypted_char)
                key_index += 1
            else:
                decrypted_text.append(ch)
        return ''.join(decrypted_text)

    def railFenceDecipher(self, encryptedText: str, rails: int) -> str:
        n = len(encryptedText)
        # Create fence matrix filled with null character (as in Java)
        fence = [['\0' for _ in range(n)] for _ in range(rails)]

        direction = -1
        row = 0
        col = 0

        # Mark positions with '*'
        for i in range(n):
            if row == 0 or row == rails - 1:
                direction = -direction
            fence[row][col] = '*'
            col += 1
            row += direction

        # Fill in the ciphertext in row-major order (only at '*')
        idx = 0
        for i in range(rails):
            for j in range(n):
                if fence[i][j] == '*':
                    fence[i][j] = encryptedText[idx]
                    idx += 1

        # Read the result in zigzag order
        plaintext = []
        direction = -1
        row = 0
        col = 0
        for i in range(n):
            if row == 0 or row == rails - 1:
                direction = -direction
            plaintext.append(fence[row][col])
            col += 1
            row += direction

        return ''.join(plaintext)

File: data/derived_cot/rq2_eval/translation/r1/_converted_code/py_to_cpp/DecryptionUtils.cpp
#include <string>
#include <vector>
#include <cctype>
#include <stdexcept>

class DecryptionUtils {
private:
    std::string key;

public:
    DecryptionUtils(std::string key) : key(key) {}

    std::string caesar_decipher(std::string ciphertext, int shift) {
        std::string plaintext;
        for (char c : ciphertext) {
            if (std::isalpha(c)) {
                char base = std::isupper(c) ? 'A' : 'a';
                int shifted = (c - base - shift) % 26;
                if (shifted < 0) {
                    shifted += 26;
                }
                plaintext += static_cast<char>(base + shifted);
            } else {
                plaintext += c;
            }
        }
        return plaintext;
    }

    std::string vigenere_decipher(std::string ciphertext) {
        if (key.empty()) {
            throw std::invalid_argument("Key must not be empty");
        }

        std::string decrypted_text;
        int key_index = 0;
        for (char c : ciphertext) {
            if (std::isalpha(c)) {
                char key_char = std::tolower(key[key_index % key.length()]);
                int shift = key_char - 'a';
                char lower_c = std::tolower(c);
                int shifted = (lower_c - 'a' - shift) % 26;
                if (shifted < 0) {
                    shifted += 26;
                }
                char decrypted_char = 'a' + shifted;
                if (std::isupper(c)) {
                    decrypted_char = std::toupper(decrypted_char);
                }
                decrypted_text += decrypted_char;
                key_index++;
            } else {
                decrypted_text += c;
            }
        }
        return decrypted_text;
    }

    std::string rail_fence_decipher(std::string encrypted_text, int rails) {
        if (rails <= 0) {
            return "";
        }
        int n = encrypted_text.length();
        if (rails == 1 || n == 0) {
            return encrypted_text;
        }

        std::vector<int> lengths(rails, 0);
        int row = 0;
        int dir = -1;
        for (int i = 0; i < n; i++) {
            lengths[row]++;
            if (row == 0 || row == rails - 1) {
                dir = -dir;
            }
            row += dir;
        }

        std::vector<std::string> rails_str;
        int start = 0;
        for (int i = 0; i < rails; i++) {
            rails_str.push_back(encrypted_text.substr(start, lengths[i]));
            start += lengths[i];
        }

        std::vector<int> ptr(rails, 0);
        row = 0;
        dir = -1;
        std::string plain_text;
        for (int i = 0; i < n; i++) {
            plain_text += rails_str[row][ptr[row]];
            ptr[row]++;
            if (row == 0 || row == rails - 1) {
                dir = -dir;
            }
            row += dir;
        }

        return plain_text;
    }
};